bloxid/v0.go

```diff
@@ -0,0 +1,131 @@
+package bloxid
+
+import (
+	"crypto/rand"
+	"encoding/base32"
+	"errors"
+	"fmt"
+	"strings"
+)
+
+const (
+	VersionUnknown Version = iota
+	Version0       Version = iota
+)
+
+type Version uint8
+
+func (v Version) String() string {
+	name, ok := version_name[v]
+	if !ok {
+		return version_name[VersionUnknown]
+	}
+	return name
+}
+
+var version_name = map[Version]string{
+	VersionUnknown: "unknown",
+	Version0:       "blox0",
+}
+
+var name_version = map[string]Version{
+	"unknown": VersionUnknown,
+	"blox0":   Version0,
+}
+
+var (
+	ErrEmptyBloxID error = errors.New("empty bloxid")
+)
+
+func NewV0(bloxid string) (*V0, error) {
+	if len(bloxid) == 0 {
+		return nil, ErrEmptyBloxID
+	}
+	return nil, nil
+}
+
+type ID interface {
+	String() string
+	ShortID() string
+	Version() string // V0
+	Type() string    // "host"
+	Region() string  // "us-east-1"
+}
+
+type V0 struct {
+	version      Version
+	region       string
+	customSuffix string
+	decoded      string
+	encoded      string
+	entityType   string
+}
+
+func (v *V0) String() string {
+	s := []string{
+		v.Version(),
+		v.entityType,
+		v.region,
+		v.encoded,
+	}
+	return strings.Join(s, ".")
+}
+
+func (v *V0) Version() string {
+	return v.version.String()
+}
+
+type V0Options struct {
+	Region     string
+	EntityType string
+	shortid    string
+}
+
+type GenerateV0Opts func(o *V0Options)
+
+func WithShortID(shortid string) func(o *V0Options) {
+	return func(o *V0Options) {
+		o.shortid = shortid
+	}
+}
+
+func GenerateV0(opts *V0Options, fnOpts ...GenerateV0Opts) (*V0, error) {
+
+	for _, fn := range fnOpts {
+		fn(opts)
+	}
+
+	encoded, decoded := uniqueID(opts)
+
+	return &V0{
+		version:    Version0,
+		region:     opts.Region,
+		decoded:    decoded,
+		encoded:    encoded,
+		entityType: opts.EntityType,
+	}, nil
+}
+
+func uniqueID(opts *V0Options) (encoded string, decoded string) {
+	entropy := randDefault()
+	decoded = fmt.Sprintf("%x", entropy)
+	encoded = strings.ToLower(base32.StdEncoding.EncodeToString(entropy))
+	return
+}
+
+const (
+	DefaultEntropySize = 40
+)
+
+func randDefault() []byte {
+	return randBytes(DefaultEntropySize)
+}
+
+func randBytes(size int) []byte {
+	b := make([]byte, size/2)
+	_, err := rand.Read(b)
+	if err != nil {
+		panic(fmt.Sprintf("unable to read /dev/urandom: %s", err))
```

Comment:
It is better to have a more generic error here like `fail to generate random value: %s`. `unable to read /dev/urandom` may not be valid for some platforms.

Reply:
The error out of std library will clarify error, the method generally returns results from /dev/urandom but here's the verbose version: https://github.com/golang/go/blob/master/src/crypto/rand/rand.go#L14-L18

getrandom(2) http://man7.org/linux/man-pages/man2/getrandom.2.html

Reply:
This code is in open source, so we should consider all platforms. For example if code is executed [On Windows systems, Reader uses the CryptGenRandom API](https://github.com/golang/go/blob/master/src/crypto/rand/rand.go#L17). In this case seeing `/dev/urandom` in error do not make any sense.  This is why we should avoid this and make message more generic. 
It is better even to just `panic(err)` here, because referring to `/dev/urandom` is not always valid.